fibre/client.go

```
@@ -0,0 +1,130 @@
+	}
+}
+
+// Client is the Fibre DA client.
+type Client struct {
+	cfg ClientConfig
+
+	txClient *user.TxClient
+	keyring  keyring.Keyring
+	valGet   validator.SetGetter
+	hostReg  validator.HostRegistry
+
+	log    *slog.Logger
+	tracer trace.Tracer
+	clock  clock.Clock
+
+	clientCache *grpc.ClientCache
+	uploadSem   chan struct{}
+	downloadSem chan struct{}
+
+	// closeWg tracks ongoing Put/Get operations and their spawned goroutines.
+	// Close() waits for this WaitGroup to ensure all operations complete before releasing resources.
+	// Put/Get operations don't wait for their spawned goroutines, allowing them to return early for low latency.
+	closeWg sync.WaitGroup
+}
+
+// NewClient creates a new [Client] with the provided dependencies.
+func NewClient(txClient *user.TxClient, kr keyring.Keyring, valGet validator.SetGetter, hostReg validator.HostRegistry, cfg ClientConfig) *Client {
+	if cfg.NewClientFn == nil {
+		cfg.NewClientFn = grpc.DefaultNewClientFn(hostReg)
+	}
+	if cfg.Tracer == nil {
+		cfg.Tracer = otel.Tracer("fibre")
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.Default()
+	}
+	if cfg.Clock == nil {
+		cfg.Clock = clock.New()
+	}
+
+	return &Client{
+		cfg:         cfg,
+		txClient:    txClient,
+		keyring:     kr,
+		valGet:      valGet,
+		hostReg:     hostReg,
+		log:         cfg.Log,
+		tracer:      cfg.Tracer,
+		clock:       cfg.Clock,
+		clientCache: grpc.NewClientCache(cfg.NewClientFn, cfg.UploadConcurrency),
+		uploadSem:   make(chan struct{}, cfg.UploadConcurrency),
+		downloadSem: make(chan struct{}, cfg.DownloadConcurrency),
+	}
+}
+
+// Config returns the [ClientConfig] used by this client.
+func (c *Client) Config() ClientConfig {
+	return c.cfg
+}
+
+// Close closes the client and releases any associated resources.
+// It waits for all ongoing Put/Get operations to complete before closing.
+func (c *Client) Close() error {
+	c.closeWg.Wait()
+	return c.clientCache.Close()
+}
```

If we want graceful shutdown Close() should also prevent allowing of new Puts/Gets. Current logic does not guarantee that, because new woperation can be created after wg is done and before clientCache is closed. non-blocking: Can be solved as followup PR

---

Yes, we could return an error saying the client is closed